### ISSUE TITLE
weird HTTP::Client issue

### ISSUE THREAD

**Issue by stakach** [crystal-lang/crystal#7383]
using crystal lang 0.27.2 on OSX
Have a weird issue with the following code:

```crystal
require "http/client"
context = OpenSSL::SSL::Context::Client.new
context.verify_mode = OpenSSL::SSL::VerifyMode::NONE
response = HTTP::Client.get "https://www.overclockers.com.au/", tls: context
response.status_code
```

It results in

```
Unhandled exception: SSL_read: Unexpected EOF (OpenSSL::SSL::Error)
  from /usr/local/Cellar/crystal/0.27.2/src/openssl/ssl/socket.cr:110:9 in 'unbuffered_read'
  from /usr/local/Cellar/crystal/0.27.2/src/io/buffered.cr:227:5 in 'fill_buffer'
  from /usr/local/Cellar/crystal/0.27.2/src/io/buffered.cr:38:5 in 'read_byte'
  from /usr/local/Cellar/crystal/0.27.2/src/http/content.cr:69:7 in 'read_byte'
  from /usr/local/Cellar/crystal/0.27.2/src/gzip/reader.cr:105:22 in 'unbuffered_read'
  from /usr/local/Cellar/crystal/0.27.2/src/io/buffered.cr:67:16 in 'read'
  from /usr/local/Cellar/crystal/0.27.2/src/io/encoding.cr:76:45 in 'read'
  from /usr/local/Cellar/crystal/0.27.2/src/io.cr:548:11 in 'gets_to_end'
  from /usr/local/Cellar/crystal/0.27.2/src/http/client/response.cr:72:15 in 'consume_body_io'
  from /usr/local/Cellar/crystal/0.27.2/src/http/client/response.cr:96:9 in 'from_io?'
  from /usr/local/Cellar/crystal/0.27.2/src/http/client.cr:512:5 in 'exec_internal_single'
  from /usr/local/Cellar/crystal/0.27.2/src/http/client.cr:498:5 in 'exec_internal'
  from /usr/local/Cellar/crystal/0.27.2/src/http/client.cr:494:5 in 'exec'
  from /usr/local/Cellar/crystal/0.27.2/src/http/client.cr:612:5 in 'exec'
  from /usr/local/Cellar/crystal/0.27.2/src/http/client.cr:640:7 in 'exec'
  from /usr/local/Cellar/crystal/0.27.2/src/http/client.cr:342:3 in 'get:tls'
```

putting almost any other website works as expected - so this feels like a new edge case as the website works in a browser fine.

**Comment by sdogruyol** [comment]
I've also reproduced this on `Arch Linux 4.20.5`

**Comment by mamantoha** [comment]
That's strange because the next code works fine:

```crystal
require "http/client"

url = "https://www.overclockers.com.au"
uri = URI.parse(url)

client = HTTP::Client.new(uri)
response = client.get("/")
puts response.status_code
```

**Comment by ysbaddaden** [comment]
No need to use a custom SSL context. The following fails just the same:
```crystal
require "http/client"
response = HTTP::Client.get "https://www.overclockers.com.au/"
puts response.status_code
```

**Comment by r00ster91** [comment]
It seems @omarroth found a fix for this issue. Can this maybe be applied to the standard library somehow?

**Comment by bcardiff** [comment]
If someone can make an explanation from that diff and send a PR it will be actionable. Thanks GH that advertise those fixes, right?

**Comment by r00ster91** [comment]
Yeah, luckily GitHub displays the commits here that reference this issue.

**Comment by omarroth** [comment]
For anyone interested in investigating further, the issue as @stakach mentioned is that the remote hangs up immediately after sending the response.

The two changes [here](https://github.com/omarroth/invidious/blob/0cf187dee73ede397c1ef67a61133d9736738fd5/src/invidious/helpers/handlers.cr#L216-L240) check for any remaining data in the SSL socket, and if there is none, return (or don't attempt to close the socket).

AFAICT this will work in all cases where this is an issue, but the "fix" can fail in normal usage (currently it's only applied when using sites where I encounter this issue).

I expect there's a cleaner fix to be made somewhere in [here](https://github.com/crystal-lang/crystal/blob/7dae2e712d2db9ba9c22586fe2308156617b9b5e/src/openssl/ssl/socket.cr) but unfortunately haven't had much time to look into it myself.

**Comment by rdp** [comment]
So with HTTP unspecified content length it should just wait until the server sends EOF?  For some reason with the failing initial example it doesn't use "Transfer-Encoding" => "chunked" but it does for the successful one.  Weird, must be sending in different headers or something.  So in the failing instance it has to just read to the end of stream which it doesn't do well with SSL sockets yet.

My hunch here is what we have is a "misbehaving" SSL server that is just closing the socket instead of gracefully sending the SSL shutdown message, then closing it. 
here's the trace:
```
Unhandled exception: SSL_read: Unexpected EOF (OpenSSL::SSL::Error)
  from src/openssl/ssl/socket.cr:0:9 in 'unbuffered_read'
  from src/io/buffered.cr:214:12 in 'fill_buffer'
  from src/io/buffered.cr:50:5 in 'read_byte'
  from src/http/content.cr:68:7 in 'read_byte'
  from src/gzip/reader.cr:105:22 in 'unbuffered_read'
  from src/io/buffered.cr:79:16 in 'read'
  from src/io/encoding.cr:76:45 in 'read'
  from src/io.cr:548:11 in 'gets_to_end'
  from src/http/client/response.cr:81:15 in 'consume_body_io'
  from src/http/client/response.cr:106:9 in 'from_io?'
  from src/http/client.cr:581:5 in 'exec_internal_single'
  from src/http/client.cr:567:5 in 'exec_internal'
  from src/http/client.cr:563:5 in 'exec'
  from src/http/client.cr:685:5 in 'exec'
  from src/http/client.cr:717:7 in 'exec'
  from src/http/client.cr:389:3 in 'get'
  from test5.cr:2:1 in '__crystal_main'
  from src/crystal/main.cr:97:5 in 'main_user_code'
  from src/crystal/main.cr:86:7 in 'main'
  from src/crystal/main.cr:106:3 in 'main'
  from __libc_start_main
  from _start
  from ???
```

 How to handle it though...
Here's one way:

```
diff --git a/src/openssl/ssl/socket.cr b/src/openssl/ssl/socket.cr
index 7a843045d..3faa0a2c8 100644
--- a/src/openssl/ssl/socket.cr
+++ b/src/openssl/ssl/socket.cr
@@ -115,7 +115,11 @@ abstract class OpenSSL::SSL::Socket < IO
 
     LibSSL.ssl_read(@ssl, slice.to_unsafe, count).tap do |bytes|
       if bytes <= 0 && !LibSSL.ssl_get_error(@ssl, bytes).zero_return?
-        raise OpenSSL::SSL::Error.new(@ssl, bytes, "SSL_read")
+        ex = OpenSSL::SSL::Error.new(@ssl, bytes, "SSL_read")
+        if ex.message && ex.message.not_nil!.includes?("Unexpected EOF")
+          return 0 # it's similar to an EOF, connection was closed without gracefully closing SSL #7383...
+        end
+        raise ex
       end
     end
   end
```

**Comment by rdp** [comment]
I also wonder how Ruby handles this (or anybody else that uses OpenSSL too...). Appears that the official way changed with 1.1.0c yikes: https://github.com/openssl/openssl/issues/1903 don't know what that means, the patch above works for me in both Linux and OS X FWIW...

**Comment by straight-shoota** [comment]
Unexpected EOF in a SSL stream is an error, so technically it should raise. But the application using the SSL stream (for example `HTTP::Client`) might decide that this is okay.

Having some insight on behaviour of other implementations would help.

**Comment by stakach** [comment]
Ruby EventMachine uses connection close when there is no content-length header: https://github.com/eventmachine/eventmachine/blob/master/lib/em/protocols/httpclient.rb#L197

Which is also how all the mainstream browsers function too (otherwise these poorly functioning server implementations would not exist)

**Comment by rdp** [comment]
Looks like ruby just treats it as an "EOF": https://github.com/ruby/ruby/blob/06a04a1aa3fbf9132c61f4ced9582c36c96d3f65/ext/openssl/ossl_ssl.c#L1871

**Comment by rdp** [comment]
possibly this commit was supposed to already address this hmm b1abae7b7386f58761ac83e3af70620c15e0cb74 investigating...

**Comment by rdp** [comment]
Seems fixed by #8540 recommend close, thanks for the report!

**Comment by stakach** [comment]
Yep, agreed, thanks all!
